### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by uk4z**
I added e2e frontend tests

### CODE REVIEW

**Review by uk4z — COMMENTED** [review]
lgtm

**Review by uk4z — COMMENTED** [review]
lgtm

**Comment by uk4z** [comment]
So in order to make tests using playwright library, I used this repo: https://github.com/dfinity/internet-identity-playwright
to integrate II in the tests. Also I had to use a specific wasm to have a compatible II canister wasm.  I added selectors (title = "..." in svelte components) to mock user interactions. Also in the utils.ts file, I created among others,  functions to transfer tokens to the user in order to use the app.

File: src/lib/index.ts [hunk 1]
```diff
@@ -203,11 +183,7 @@ export function displayTimeLeft(created_at: number, isMobile = false) {
 	const daysLeft = Math.floor(timeLeft / 60 / 60 / 24);
 	const hoursLeft = Math.floor((timeLeft - daysLeft * 60 * 60 * 24) / 60 / 60);
 
-	if (isMobile && daysLeft > 0) {
-		return `${daysLeft} days left`;
-	} else if (isMobile && hoursLeft > 0) {
-		return `${hoursLeft} hours left`;
-	} else if (daysLeft > 0 && hoursLeft > 0) {
```
Comment (leo-ship-it) [inline src/lib/index.ts:210]: Why do you want to change this? It doesn't seem related to this PR.

File: tests/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# e2e Testing
+
+Here is a brief overview of how to setup the testing environment before running the tests:
+
+We are using a mocking minting account that will act as the intermediary between Ledger minting accounts and the supply of tokens of the Internet Identity accounts created withing the tests.
```
Comment (leo-ship-it) [inline tests/README.md:5]:
```suggestion
An `intermediary` account is responsible for distributing ICP/nICP to the principals used in this test.
```

File: tests/utils.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+const key = [
+	'302a300506032b657003210093d488f46b485c07e09b554d9451574bfc669912b99d453722c474e6a7f90fcc',
+	'90252a6913658dbb4b36b276410216d47a1891280493cd485328279a12a53e2c'
+];
+
+const parsedKey = JSON.stringify(key);
+
+// AccountId = 90526bdfd692793cba1f96bde9079994ce4d40033746f04c12064ea599e2c274
+// Principal = syna7-6ipnd-myx4g-ia46u-nxwok-u5nrr-yxgpi-iang7-lvru2-i7n23-tqe
```
Comment (leo-ship-it) [inline tests/utils.ts:25]: ?

File: tests/e2e.spec.ts [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+import { user, canisters } from '$lib/stores';
+import { get } from 'svelte/store';
+import { mockSetup, supplyICP, swap, isToastSuccess, supplyNICP, send } from './utils';
+
+const VALID_PRINCIPAL = 'l72el-pt5ry-lmj66-3opyw-tl5xx-3wzfl-n3mja-dqirc-oxmqs-uxqe6-6qe';
+const ACCOUNT_ID = 'e73a99617af2a8dbfe9b75e463e83a905e30aa50250972ad19c21922c22b2a2a';
+const VALID_ACCOUNT =
+	'bd3sg-teaaa-aaaaa-qaaba-cai-kdmvzla.5f0e93000f4cbd9db8c36d27cad8b8a97706c0710154172029e54541e18fd180';
+
+test('Mock minting account has balance', async () => {
```
Comment (leo-ship-it) [inline tests/e2e.spec.ts:12]:
```suggestion
test('Account responsible for token dispatch should have balance', async () => {
```

File: tests/e2e.spec.ts [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+	await page.goto('/any');
+	await expect(page).toHaveURL('/stake/');
+
+	await page.goto('/sns/');
+	await expect(page).toHaveURL('/sns/');
+
+	await page.goto('/wallet/');
+	await expect(page).toHaveURL('/stake/');
+
+	await page.locator('[title="connect-btn"]').click();
```
Comment (leo-ship-it) [inline tests/e2e.spec.ts:59]:
```suggestion
      // Check redirections
	await page.goto('/');
	await expect(page).toHaveURL('/stake/');

	await page.goto('/any');
	await expect(page).toHaveURL('/stake/');

	await page.goto('/sns/');
	await expect(page).toHaveURL('/sns/');

	await page.goto('/wallet/');
	await expect(page).toHaveURL('/stake/');

	await page.locator('[title="connect-btn"]').click();
```

File: tests/e2e.spec.ts [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+	await iiPage.signInWithNewIdentity({ selector: '[title="ii-connect-btn"]' });
+
+	await page.locator('#disconnect-btn').click();
```
Comment (leo-ship-it) [inline tests/e2e.spec.ts:63]:
```suggestion
      // Should login and logout
	await iiPage.signInWithNewIdentity({ selector: '[title="ii-connect-btn"]' });

	await page.locator('#disconnect-btn').click();
```

File: tests/utils.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+		const actors = await fetchActors(agent);
+
+		canisters.set(new Canisters(actors));
+		user.set(new User(dummyIdentity.getPrincipal()));
+	} catch (error) {
+		console.error('Login failed:', error);
+	}
+};
+
+export async function supplyICP(accountString: string) {
```
Comment (leo-ship-it) [inline tests/utils.ts:44]:
```suggestion
export async function transferICP(accountString: string) {
```

File: tests/e2e.spec.ts [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+
+	await walletInfo.click();
+
+	const accountId = await page
+		.locator('p[title="accountIdentifier-hex"]')
+		.evaluate((accountId) => accountId.textContent);
+
+	if (!accountId) throw new Error('No account id found.');
+
+	const paragraphs = walletInfo.locator('p');
```
Comment (leo-ship-it) [inline tests/e2e.spec.ts:84]: brittle add title or id to select directly the desired element.

File: tests/e2e.spec.ts [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+	if (!accountId) throw new Error('No account id found.');
+
+	const principal = await page
+		.locator('p[title="principal-user"]')
+		.evaluate((principal) => principal.textContent);
+
+	if (!principal) throw new Error('No principal found.');
+
+	await supplyICP(accountId);
+	await supplyNICP(principal);
```
Comment (leo-ship-it) [inline tests/e2e.spec.ts:183]:
```suggestion
	await transfernICP(principal);
```

File: tests/e2e.spec.ts [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+
+	const paragraphs = walletInfo.locator('p');
+	expect(await paragraphs.count()).toEqual(3);
+	await expect(paragraphs.nth(0)).toHaveText('15 ICP');
+	await expect(paragraphs.nth(1)).toHaveText('15 nICP');
```
Comment (leo-ship-it) [inline tests/e2e.spec.ts:188]: same comment about using locator p tag

File: tests/e2e.spec.ts [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+
+	await page.locator('[title="send-btn-ICP"]').click();
+
+	await send(page, 'aaa-aa', '10');
```
Comment (leo-ship-it) [inline tests/e2e.spec.ts:192]:
```suggestion
	await send(page, 'aaa-aa', '10'); // Principal invalid, valid anonymous principal is aaaa-aa
```

File: tests/e2e.spec.ts [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+	await expect(page.locator('span.error')).toBeVisible();
+	await send(page, ACCOUNT_ID, '0.000000009');
```
Comment (leo-ship-it) [inline tests/e2e.spec.ts:194]: Cannot give more than 8 decimals.

File: tests/e2e.spec.ts [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+	expect(await isToastSuccess(page)).toBeTruthy();
+});
+
+// This test is expected to work only with one worker (otherwise one test impact the others).
+// Use npx playwright test --prokect=chromium (you can use firefox or webkit too).
+test('e2e test sns', async ({ page }) => {
+	await page.goto('/sns');
+	await expect(page.locator('.sns-listing')).toBeVisible();
+	const snsList = page.locator('.sns-listing').locator('div');
+	expect(await snsList.count()).toEqual(21);
```
Comment (leo-ship-it) [inline tests/e2e.spec.ts:247]: https://testing.googleblog.com/2015/01/testing-on-toilet-change-detector-tests.html

File: package.json [hunk 14]
```diff
@@ -8,12 +8,15 @@
 		"preview": "vite preview",
 		"check": "svelte-kit sync && svelte-check --tsconfig ./tsconfig.json",
 		"check:watch": "svelte-kit sync && svelte-check --tsconfig ./tsconfig.json --watch",
-		"test": "vitest",
```
[the remark this answers is not in the document]
Comment (uk4z) [reply package.json:11]: nope. Available in the lib.spec.ts file

File: playwright.config.ts [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+		/* Test against branded browsers. */
+		// {
+		//   name: 'Microsoft Edge',
+		//   use: { ...devices['Desktop Edge'], channel: 'msedge' },
+		// },
+		// {
+		//   name: 'Google Chrome',
+		//   use: { ...devices['Desktop Chrome'], channel: 'chrome' },
+		// },
```
[the remark this answers is not in the document]
Comment (uk4z) [reply playwright.config.ts:73]: Not for now because it would require a lot of adjustments. But at some point we should add them probably

File: src/lib/index.ts [hunk 16]
```diff
@@ -149,26 +149,6 @@ export function computeRewards(alreadyDistributed: BigNumber, converting: BigNum
 	return totalRewards;
 }
 
-export function nicpLeftUntilNextTier(alreadyDistributed: BigNumber): {
```
[the remark this answers is not in the document]
Comment (uk4z) [reply src/lib/index.ts:152]: This function: computeRewards

[the remark this answers is not in the document]
Comment (uk4z) [reply src/lib/index.ts:152]: Actually I made this function back then to display the number of tokens left to be minted before the halving of WTN

File: src/lib/index.ts (hunk 1 above)
Comment (uk4z) [reply src/lib/index.ts:210]: Oh I missed it, I would have put it in the fix PR.

File: tests/README.md (hunk 2 above)
Comment (uk4z) [reply tests/README.md:5]: OKay

File: playwright.config.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (leo-ship-it) [reply playwright.config.ts:73]: Let's do a follow up PR.

**Review by uk4z — COMMENTED** [review]
lgtm
